README.md
# Surfer

Surfer is a Simple static file server.
It comes with a commandline tool to upload files from your local folders and a webinterface to manage files directly on the server.

## Installation

[![Install](https://cloudron.io/img/button32.png)](https://cloudron.io/button.html?app=io.cloudron.surfer)

or using the [Cloudron command line tooling](https://docs.cloudron.io/packaging/cli/)

```bash
cloudron install --appstore-id io.cloudron.surfer
```

## Building

### Cloudron

The app package can be built using the [Cloudron command line tooling](https://docs.cloudron.io/packaging/cli/).

```bash
git clone https://git.cloudron.io/cloudron/surfer.git
cd surfer
cloudron build
cloudron install
```

### Standalone

Surfer can also be run standlone on any server:

```bash
git clone https://git.cloudron.io/cloudron/surfer.git
cd surfer
npm install
npm run build
./server.js <public folder>
```

Use the `admin` tool to manage local users.

## File management

The admin interface is available under the `/_admin` location or you can upload files using the commandline tool.

First, install the surfer cli tool using npm.

```bash
npm -g install cloudron-surfer
```

Login using your Cloudron credentials:

```bash
surfer login <this app's url>
```

Put some files:

```bash
surfer put [file]
```

## Testing

The e2e tests are located in the `test/` folder and require [nodejs](http://nodejs.org/). They are creating a fresh build, install the app on your Cloudron, perform tests, backup, restore and test if the files are still ok.

```bash
cd surfer

npm install
USERNAME=<cloudron username> PASSWORD=<cloudron password> mocha --bail test/test.js
```

## Development

```bash
git clone https://git.cloudron.io/cloudron/surfer.git
cd surfer
npm install
```

During UI development, the assets have to be built after changes. This can be done automatically with

```bash
npm run watch
```

While having the `./server.js` instance running.


CHANGELOG
[0.4.0]
* Initial version

[0.4.1]
* Remove silly mediaLinks

[0.4.2]
* Fix app url

[0.4.3]
* Improve healthcheck

[0.5.0]
* Add settings ui

[0.5.1]
* Fix settings link
* Add selenium tests

[0.5.2]
* Support username and email login

[0.5.3]
* Support uploading the same file multiple times

[0.5.4]
* Fix upload files first time

[0.5.5]
* Use new base image

[0.5.6]
* Use new base image 0.9.0

[0.5.7]
* Add post install message
* Download files as attachments

[0.5.8]
* Various encoding bugfixes
* Make breadcumbs clickable

[0.6.0]
* Support multifile upload via the webinterface

[4.0.0]
* Implement token auth with redis backend
* Basic drag'n'drop support
* Use new base image v0.10.0
* Synchronize cli and server versions

[4.1.0]
* Add ability to download files from the UI

[4.1.1]
* Add directory listing sorting
* Replace github links with our own gitlab

[5.0.0]
* Allow directory listing to be configurable
* Refactor the UI

[5.1.0]
* Add directory upload support
* Improve drag'n'drop of files

[5.2.0]
* Disable folder listing by default
* Do not render welcome screen if folder listing is enabled

[5.3.0]
* Replacing redis token store with local json file
* Add upload progress reporting
* Minor interface improvements

[5.4.0]
* Use new Cloudron base image
* Send 404 status code correctly

[5.5.0]
* Add webdav endpoint support
* Use inline edit for rename
* Various interface layout fixes

[5.6.0]
* Add SFTP support

[5.7.0]
* Update to latest app package manifest

[5.8.0]
* Add access token support for better CI/CD integration
* Various UI fixes for preview thumbnails

[5.9.0]
* Rename `surfer_root` to `public`
* `surfer put` CLI has changed

[5.9.1]
* Visual improvements to give more space to content
* Add preview sidebar
* Update dependencies for securty fixes

[5.9.2]
* Improve public folder listing

[5.10.0]
* Update to new Cloudron base image

[5.10.1]
* Add breadcrumbs to public folder listing
* Minor UI improvements

[5.10.2]
* Fix public folder listing if path contains spaces #14
* Improve breadcrumb layout
* Fix chrome bug on file rename #15

[5.10.3]
* Sanitize username input to prevent LDAP DoS attack. Thanks to Alessio (d3lla) for reporting!

[5.10.4]
* Update screenshots
* (security) Use crypto.random instead of uuid to generate API tokens
* API tokens are now base64 safe (to make it suitable for masked GitLab CI variables)

[5.11.0]
* Improve file uploading through cli
* Purge remote files during surfer put with --delete flag
* Improve UI login behavior and autofocus

[5.11.1]
* Fix file uploading bug

[5.11.2]
* Allow ability to remove all files via cli
* Update all dependencies

[5.11.3]
* Fix public viewing of subfolders
* Add music type icon

[5.12.0]
* Add CORS support

[5.12.1]
* Fix extensive WebDAV usage by caching auth

[5.12.2]
* Fix crash in cli tool

[5.13.0]
* Rewritten UI in primevue
* Add support for listing folders first

[5.13.1]
* Rewritten UI in primevue
* Add support for listing folders first
* Ability to specify site title and favicon

[5.13.2]
* Add ability to password or user protect the site
* Update node modules

[5.13.3]
* Fix previews when site is protected

[5.13.4]
* Fix case where access restriction is on and public folder listing off

[5.13.5]
* Add file name to preview
* Allow to close preview
* Fix dialog closing when clicking on backdrop
* Update dependencies

[5.14.0]
* Enable multi-domain flag for setting domain aliases

[5.14.1]
* Fix upload progress

[5.15.0]
* Allow to have custom 404.htm(l) in root folder
* Fix back button
* Update to base image v3

[5.16.0]
* Add basic support for noscript usage
* Improve preview behavior
* Be more conform with web usage and create proper web links and allow initial selection to prepare multiselect
* Fix alphanumeric sorting

[5.16.1]
* Use proper mime-type database for content-type delivery
* Fix rename selection bug
* Allow keyboard up/down navigation

[5.16.2]
* Allow to set index filename

[5.16.3]
* Fixup various path encoding issues

[5.16.4]
* Fix public listing regression for folders with special characters

[5.16.5]
* Use secure cookies

[5.16.6]
* Fix action overflow in file listing

[5.16.7]
* Update base image to v3.2.0

[5.17.0]
* Update dependencies
* Reduce flickering on login
* Reduce padding in list

[5.17.1]
* Migrate to upstreamVerison in manifest
* Update dependencies

[5.17.2]
* Always protect /api/files/*
* Stop using GET /api/files/* in UI

[5.17.3]
* Send 401 status code instead of 200 when authentication is required first
* Update dependencies

[5.17.4]
* Update dependencies
* Only show preview for supported types to avoid immediate download
* Fix various vuejs warnings

[5.17.5]
* Update depenencies
* Fixup cli tool to report better error status codes

[5.17.6]
* Further cli tool exit status code fixes



server.js
#!/usr/bin/env node

'use strict';

var express = require('express'),
    async = require('async'),
    morgan = require('morgan'),
    path = require('path'),
    session = require('express-session'),
    ejs = require('ejs'),
    fs = require('fs'),
    crypto = require('crypto'),
    cors = require('./src/cors.js'),
    copyFile = require('./src/copyFile.js'),
    compression = require('compression'),
    contentDisposition = require('content-disposition'),
    bodyParser = require('body-parser'),
    lastMile = require('connect-lastmile'),
    HttpError = require('connect-lastmile').HttpError,
    HttpSuccess = require('connect-lastmile').HttpSuccess,
    multipart = require('./src/multipart'),
    auth = require('./src/auth.js'),
    mime = require('./src/mime.js'),
    webdav = require('webdav-server').v2,
    files = require('./src/files.js')(path.resolve(__dirname, process.argv[2] || 'files'));

const ROOT_FOLDER = path.resolve(__dirname, process.argv[2] || 'files');
const CONFIG_FILE = path.resolve(__dirname, process.argv[3] || '.config.json');
const FAVICON_FILE = path.resolve(__dirname, process.argv[4] || 'favicon.png');
const FAVICON_FALLBACK_FILE = path.resolve(__dirname, 'dist', 'logo.png');

const PASSWORD_PLACEHOLDER = '__PLACEHOLDER__';

const CRYPTO_SALT_SIZE = 64; // 512-bit salt
const CRYPTO_ITERATIONS = 10000; // iterations
const CRYPTO_KEY_LENGTH = 512; // bits
const CRYPTO_DIGEST = 'sha1'; // used to be the default in node 4.1.1 cannot change since it will affect existing db records

var sessionStore = new session.MemoryStore();

// Ensure the root folder exists
fs.mkdirSync(ROOT_FOLDER, { recursive: true });

var config = {
    folderListingEnabled: false,
    sortFoldersFirst: true,
    title: '',
    accessRestriction: '',
    accessPassword: ''
};

function setServMiddlewareHeaders (res, path) {
    // handle ?download in query
    if ('download' in res.req.query) res.setHeader('Content-Disposition', contentDisposition(path));
}

// we will regenerate this if settings change
var staticServMiddleware = express.static(ROOT_FOLDER, { index: 'index.html', setHeaders: setServMiddlewareHeaders });

function getSettings(req, res) {
    res.send({
        folderListingEnabled: !!config.folderListingEnabled,
        sortFoldersFirst: !!config.sortFoldersFirst,
        title: config.title || 'Surfer',
        index: config.index || '',
        accessRestriction: config.accessRestriction || '',
        accessPassword: config.accessPassword ? PASSWORD_PLACEHOLDER : '' // don't send the password, helps the UI to figure if a password was set at all
    });
}

function setSettings(req, res, next) {
    if (typeof req.body.folderListingEnabled !== 'boolean') return next(new HttpError(400, 'missing folderListingEnabled boolean'));
    if (typeof req.body.sortFoldersFirst !== 'boolean') return next(new HttpError(400, 'missing sortFoldersFirst boolean'));
    if (typeof req.body.title !== 'string') return next(new HttpError(400, 'missing title string'));
    if (req.body.index && typeof req.body.index !== 'string') return next(new HttpError(400, 'index must be falsy or a string'));
    if (typeof req.body.accessRestriction !== 'string') return next(new HttpError(400, 'missing accessRestriction string'));
    if ('accessPassword' in req.body && typeof req.body.accessPassword !== 'string') return next(new HttpError(400, 'accessPassword must be a string'));

    function clearNonAdminSessions(callback) {
        callback = callback || function () {};

        sessionStore.all(function (error, sessions) {
            if (error) return console.error('Failed to get sessions.', error);

            async.each(Object.keys(sessions), function (sid, callback) {
                if (sessions[sid].isAdmin) return callback();

                sessionStore.destroy(sid, function (error) {
                    if (error) console.error(`Failed to delete session %sid.`, error);
                    callback();
                });
            }, callback);
        });
    }

    function updatePasswordIfNeeded(callback) {
        if (!('accessPassword' in req.body) || req.body.accessPassword === PASSWORD_PLACEHOLDER) return callback();

        crypto.randomBytes(CRYPTO_SALT_SIZE, function (error, salt) {
            if (error) return callback(error);

            crypto.pbkdf2(req.body.accessPassword, salt, CRYPTO_ITERATIONS, CRYPTO_KEY_LENGTH, CRYPTO_DIGEST, function (error, derivedKey) {
                if (error) return callback(error);

                config.accessPassword = Buffer.from(derivedKey, 'binary').toString('hex');
                config.accessPasswordSalt = salt.toString('hex');

                clearNonAdminSessions(callback);
            });
        });
    }

    config.folderListingEnabled = !!req.body.folderListingEnabled;
    config.sortFoldersFirst = !!req.body.sortFoldersFirst;
    config.title = req.body.title;
    config.index = req.body.index;

    staticServMiddleware = express.static(ROOT_FOLDER, { index: config.index || 'index.html', setHeaders: setServMiddlewareHeaders });

    // if changed invalidate sessions
    if (config.accessRestriction !== req.body.accessRestriction) clearNonAdminSessions();

    config.accessRestriction = req.body.accessRestriction;

    updatePasswordIfNeeded(function (error) {
        if (error) return next(new HttpError(500, 'failed to set password'));

        fs.writeFile(CONFIG_FILE, JSON.stringify(config), function (error) {
            if (error) return next(new HttpError(500, 'unable to save settings'));

            next(new HttpSuccess(201, {}));
        });
    });

}

function getFavicon(req, res) {
    if (fs.existsSync(FAVICON_FILE)) res.sendFile(FAVICON_FILE);
    else res.sendFile(FAVICON_FALLBACK_FILE);
}

function setFavicon(req, res, next) {
    if (!req.files || !req.files.file) return next(new HttpError(400, 'missing file'));

    copyFile(req.files.file.path, FAVICON_FILE, function (error) {
        if (error) {
            console.error('Failed to save favicon.', error);
            return next(new HttpError(500, 'Failed to save favicon'));
        }

        next(new HttpSuccess(201, {}));
    });
}

function resetFavicon(req, res, next) {
    fs.unlink(FAVICON_FILE, function (error) {
        if (error) {
            console.error('Failed to reset favicon.', error);
            return next(new HttpError(500, 'Failed to reset favicon'));
        }

        next(new HttpSuccess(201, {}));
    });
}

function handleProtection(req, res, next) {
    if (!config.accessRestriction) return next();
    if (req.session.isValid) return next();

    res.status(401).sendFile(path.join(__dirname, '/dist/protected.html'));
}

function protectedLogin(req, res, next) {
    function success() {
        req.session.isValid = true;
        next(new HttpSuccess(200, {}));
    }

    if (config.accessRestriction === 'password') {
        let saltBinary = Buffer.from(config.accessPasswordSalt, 'hex');
        crypto.pbkdf2(req.body.password, saltBinary, CRYPTO_ITERATIONS, CRYPTO_KEY_LENGTH, CRYPTO_DIGEST, function (error, derivedKey) {
            if (error) {
                console.log('Failed to derive key.', error);
                return next(new HttpError(500, 'internal error'));
            }

            let derivedKeyHex = Buffer.from(derivedKey, 'binary').toString('hex');
            if (derivedKeyHex !== config.accessPassword) return next(new HttpError(403, 'forbidden'));

            success();
        });
    } else if (config.accessRestriction === 'user') {
        auth.verifyUser(req.body.username, req.body.password, function (error) {
            if (error) return next(new HttpError(401, 'Invalid credentials'));
            success();
        });
    } else {
        next(new HttpError(409, 'site is not protected'));
    }
}

function send404(res) {
    // first check if /404.htm(l) exists, if so send that
    if (fs.existsSync(path.join(ROOT_FOLDER, '404.html'))) return res.status(404).sendFile(path.join(ROOT_FOLDER, '404.html'));
    if (fs.existsSync(path.join(ROOT_FOLDER, '404.htm' ))) return res.status(404).sendFile(path.join(ROOT_FOLDER, '404.htm'));

    res.status(404).sendFile(__dirname + '/dist/404.html');
}

// Load the config file
try {
    console.log(`Using config file at: ${CONFIG_FILE}`);
    config = require(CONFIG_FILE);
} catch (e) {
    if (e.code === 'MODULE_NOT_FOUND') console.log(`Config file ${CONFIG_FILE} not found`);
    else console.log(`Cannot load config file ${CONFIG_FILE}`, e);
}

if (typeof config.folderListingEnabled !== 'boolean') config.folderListingEnabled = false;
if (typeof config.sortFoldersFirst !== 'boolean') config.sortFoldersFirst = true;
if (typeof config.title !== 'string') config.title = 'Surfer';
if (typeof config.accessRestriction !== 'string') config.accessRestriction = '';
if (typeof config.accessPassword !== 'string') config.accessPassword = '';

// Setup mime-type handling
mime(express);

// Setup the express server and routes
var app = express();
var router = new express.Router();

// needed for secure cookies
app.enable('trust proxy');

var webdavServer = new webdav.WebDAVServer({
    requireAuthentification: true,
    httpAuthentication: new webdav.HTTPBasicAuthentication(new auth.WebdavUserManager(), 'Cloudron Surfer')
});

webdavServer.setFileSystem('/', new webdav.PhysicalFileSystem(ROOT_FOLDER), function (success) {
    if (!success) console.error('Failed to setup webdav server!');
});

const PUBLIC_HTML = fs.readFileSync(__dirname + '/dist/public.html', 'utf8');
const PUBLIC_NOSCRIPT_EJS = fs.readFileSync(__dirname + '/frontend/public.noscript.ejs', 'utf8');

var multipart = multipart({ maxFieldsSize: 2 * 1024, limit: '512mb', timeout: 3 * 60 * 1000 });

router.post  ('/api/protectedLogin', protectedLogin);
router.post  ('/api/login', auth.login);
router.post  ('/api/logout', auth.verify, auth.logout);
router.get   ('/api/settings', getSettings);
router.get   ('/api/favicon', getFavicon);
router.put   ('/api/favicon', auth.verify, multipart, setFavicon);
router.delete('/api/favicon', auth.verify, resetFavicon);
router.put   ('/api/settings', auth.verify, setSettings);
router.get   ('/api/tokens', auth.verify, auth.getTokens);
router.post  ('/api/tokens', auth.verify, auth.createToken);
router.delete('/api/tokens/:token', auth.verify, auth.delToken);
router.get   ('/api/profile', auth.verify, auth.getProfile);
router.get   ('/api/files/*', auth.verify, files.get);
router.post  ('/api/files/*', auth.verify, multipart, files.post);
router.put   ('/api/files/*', auth.verify, files.put);
router.delete('/api/files/*', auth.verify, files.del);

app.use('/api/healthcheck', function (req, res) { res.status(200).send(); });
app.use(morgan('dev'));
app.use(compression());
app.use(cors({ origins: [ '*' ], allowCredentials: false }));
app.use('/api', bodyParser.json());
app.use('/api', bodyParser.urlencoded({ extended: false, limit: '100mb' }));
app.use(session({ store: sessionStore, secret: 'surfin surfin', resave: false, saveUninitialized: true, cookie: { secure: !!process.env.CLOUDRON, sameSite: 'strict' } }));
app.use(router);
app.use(webdav.extensions.express('/_webdav', webdavServer));
app.use('/_admin', express.static(__dirname + '/dist'));
app.use('/', handleProtection);
app.use('/', function (req, res, next) { staticServMiddleware(req, res, next); });
app.use('/', function welcomePage(req, res, next) {
    if (config.folderListingEnabled || req.path !== '/') return next();
    res.status(200).sendFile(path.join(__dirname, '/dist/welcome.html'));
});
app.use('/', function (req, res, next) {
    if (!config.folderListingEnabled) return send404(res);

    const filePath = req.path ? decodeURIComponent(req.path) : '';
    if (!fs.existsSync(path.join(ROOT_FOLDER, filePath))) {
        console.log('hit this?', path.join(ROOT_FOLDER, filePath));
        return send404(res);
    }

    // we provision the public app with all the info so we can do static and dynamic rendering
    files.getFolderListing(filePath, function (error, result) {
        if (error) return next(error);

        var html = ejs.render(PUBLIC_NOSCRIPT_EJS, result, {});

        var out = PUBLIC_HTML;
        out = out.replace('<noscript></noscript>', `<noscript>${html}</noscript>`);
        out = out.replace('<withscript></withscript>', `<script>window.surfer = { entries: ${JSON.stringify(result.entries)}, stat: ${JSON.stringify(result.stat)} };</script>`);

        res.status(200).send(out);
    });
});
app.use(lastMile());

var server = app.listen(3000, function () {
    var host = server.address().address;
    var port = server.address().port;

    console.log(`Base path: ${ROOT_FOLDER}`);
    console.log();
    console.log(`Listening on http://${host}:${port}`);

    auth.init(config);
});
